wideq/core.py

```diff
@@ -240,166 +345,218 @@ def parse_oauth_callback(url):
 
-    url = urljoin(api_root + '/', 'member/login')
-    data = {
-        'countryCode': country,
-        'langCode': language,
-        'loginType': 'EMP',
-        'token': access_token,
+    headers = {
+        "x-lge-appkey": CLIENT_ID,
+        "x-lge-oauth-signature": sig,
+        "x-lge-oauth-date": timestamp,
+        "Accept": "application/json",
     }
-    return lgedm_post(url, data)
 
+    res = requests.post(url, headers=headers, data=data, timeout=DEFAULT_TIMEOUT)
 
-def refresh_auth(oauth_root, refresh_token):
-    """Get a new access_token using a refresh_token.
+    if res.status_code != 200:
+        raise TokenError()
 
-    May raise a `TokenError`.
+    res_data = res.json()
+    LOGGER.debug(res_data)
+
+    return res_data
+
+
+def login(oauth_url, auth_code):
+    """Get a new access_token using an authorization_code
+    
+    May raise a `tokenError`.
     """
 
-    token_url = urljoin(oauth_root, '/oauth2/token')
-    data = {
-        'grant_type': 'refresh_token',
-        'refresh_token': refresh_token,
-    }
+    out = auth_request(
+        oauth_url,
+        {
+            "code": auth_code,
+            "grant_type": "authorization_code",
+            "redirect_uri": OAUTH_REDIRECT_URI,
+        },
+    )
 
-    # The timestamp for labeling OAuth requests can be obtained
-    # through a request to the date/time endpoint:
-    # https://us.lgeapi.com/datetime
-    # But we can also just generate a timestamp.
-    timestamp = datetime.datetime.utcnow().strftime(DATE_FORMAT)
+    return out["access_token"], out["refresh_token"]
 
-    # The signature for the requests is on a string consisting of two
-    # parts: (1) a fake request URL containing the refresh token, and (2)
-    # the timestamp.
-    req_url = ('/oauth2/token?grant_type=refresh_token&refresh_token=' +
-               refresh_token)
-    sig = oauth2_signature('{}\n{}'.format(req_url, timestamp),
-                           OAUTH_SECRET_KEY)
 
-    headers = {
-        'lgemp-x-app-key': OAUTH_CLIENT_KEY,
-        'lgemp-x-signature': sig,
-        'lgemp-x-date': timestamp,
-        'Accept': 'application/json',
-    }
+def refresh_auth(oauth_root, refresh_token):
+    """Get a new access_token using a refresh_token.
 
-    with retry_session() as session:
-        res = session.post(token_url, data=data, headers=headers)
-    res_data = res.json()
+    May raise a `TokenError`.
+    """
+    out = auth_request(
+        oauth_root, {"grant_type": "refresh_token", "refresh_token": refresh_token}
+    )
 
-    if res_data['status'] != 1:
-        raise TokenError()
-    return res_data['access_token']
+    return out["access_token"]
 
 
 class Gateway(object):
-    def __init__(self, auth_base, api_root, oauth_root, country, language):
+    def __init__(self, auth_base, api_root, api2_root, country, language):
         self.auth_base = auth_base
         self.api_root = api_root
```

Comment:
Do we still need the "old" API root?

Reply:
Absolutly yes. Some device still use old API, this is an information provided by the Gateway. Depending on the device is used the api_root or the api2_root